usr/lib/common/uri.c

```diff
@@ -0,0 +1,270 @@
+    } while(0)
+
+#define FORMAT_ATTRIBUTE_ALL(sep, txt, a, t, buf)                       \
+    do {                                                                \
+        if (((a)[0].type == (t)) &&                                     \
+            ((a)[0].pValue)) {                                          \
+            p11_buffer_append_len(buf, sep, 1);                         \
+            p11_buffer_append(buf, txt);                                \
+            p11_encode(START_END((a)[0].pValue,                         \
+                                 (a)[0].ulValueLen),                    \
+                       buf);                                            \
+            *sep = P11_URI_SEP_ATTR;                                    \
+        }                                                               \
+    } while(0)
+
+#define FORMAT_ATTRIBUTE_CLASS(sep, txt, a, buf)                        \
+    do {                                                                \
+        if (((a)[0].type == CKA_CLASS) &&                               \
+            ((a)[0].pValue) &&                                          \
+            (rfc7512_get_cko((a)[0].pValue))) {                         \
+            p11_buffer_append_len(buf, sep, 1);                         \
+            p11_buffer_append(buf, txt);                                \
+            p11_buffer_append(buf, rfc7512_get_cko((a)[0].pValue));     \
+            *sep = P11_URI_SEP_ATTR;                                    \
+        }                                                               \
+    } while(0)
+
+#define FORMAT_VERSION(sep, txt, v, buf)                                \
+    do {                                                                \
+        if (((v).major != (CK_BYTE)-1) ||                               \
+            ((v).minor != (CK_BYTE)-1)) {                               \
+            p11_buffer_append_len(buf, sep, 1);                         \
+            p11_buffer_append_printf(buf, txt "%d.%d",                  \
+                                     (v).major, (v).minor);             \
+            *sep = P11_URI_SEP_ATTR;                                    \
+        }                                                               \
+    } while(0)
+
+#define FORMAT_ULONG(sep, txt, ul, buf)                                 \
+    do {                                                                \
+        if ((ul) != (CK_ULONG)-1) {                                     \
+            p11_buffer_append_len(buf, sep, 1);                         \
+            p11_buffer_append_printf(buf, txt "%lu", ul);               \
+            *sep = P11_URI_SEP_ATTR;                                    \
+        }                                                               \
+    } while(0)
+
+static const char HEX_CHARS_LOWER[] = "0123456789abcdef";
+
+static char *rfc7512_get_cko(CK_OBJECT_CLASS_PTR class)
+{
```

Comment:
Hmm, yes, is correct, but I would code it as (CK_OBJECT_CLASS *class) so that we see the *class as it is used below in the switch.

Reply:
What about using CK_PTR in the switch statement? It looks to me, that this might be the solution, how a PKCS#11-conform application/library should look like.

```
static char *rfc7512_get_cko(CK_OBJECT_CLASS_PTR class)
{
    switch (CK_PTR class) {
```

Reply:
ok, so we will go with `static char *rfc7512_get_cko(CK_OBJECT_CLASS *class)`?